server/src/com/thoughtworks/go/config/update/AddEnvironmentCommand.java

```diff
@@ -1,24 +1,82 @@
 package com.thoughtworks.go.config.update;
 
 import com.thoughtworks.go.config.*;
+import com.thoughtworks.go.config.commands.EntityConfigUpdateCommand;
+import com.thoughtworks.go.config.validation.EnvironmentAgentValidator;
+import com.thoughtworks.go.config.validation.EnvironmentPipelineValidator;
+import com.thoughtworks.go.config.validation.GoConfigValidator;
+import com.thoughtworks.go.config.validation.NameTypeValidator;
+import com.thoughtworks.go.i18n.Localizable;
+import com.thoughtworks.go.i18n.LocalizedMessage;
 import com.thoughtworks.go.server.domain.Username;
+import com.thoughtworks.go.server.service.GoConfigService;
+import com.thoughtworks.go.server.service.result.LocalizedOperationResult;
+import com.thoughtworks.go.serverhealth.HealthStateType;
 
-public class AddEnvironmentCommand implements UpdateConfigCommand, UserAware {
+import java.util.Arrays;
+import java.util.List;
+
+import static com.thoughtworks.go.config.BasicCruiseConfig.copyErrors;
+
+public class AddEnvironmentCommand implements EntityConfigUpdateCommand<EnvironmentConfig> {
+
+    private final GoConfigService goConfigService;
     private final BasicEnvironmentConfig environmentConfig;
-    private Username user;
+    private final Username user;
+    private final LocalizedOperationResult result;
+    private final List<GoConfigValidator> VALIDATORS = Arrays.asList(
+            new EnvironmentAgentValidator(),
+            new EnvironmentPipelineValidator()
+    );
 
-    public AddEnvironmentCommand(BasicEnvironmentConfig environmentConfig, Username user) {
+    public AddEnvironmentCommand(GoConfigService goConfigService, BasicEnvironmentConfig environmentConfig, Username user, LocalizedOperationResult result) {
+        this.goConfigService = goConfigService;
         this.environmentConfig = environmentConfig;
         this.user = user;
+        this.result = result;
     }
 
-    public CruiseConfig update(CruiseConfig cruiseConfig) {
-        cruiseConfig.addEnvironment(environmentConfig);
-        return cruiseConfig;
+    @Override
+    public void update(CruiseConfig preprocessedConfig) throws Exception {
+        preprocessedConfig.addEnvironment(environmentConfig);
+    }
+
+    @Override
+    public boolean isValid(CruiseConfig preprocessedConfig) {
+        for (GoConfigValidator validator : VALIDATORS) {
+            try {
+                validator.validate(preprocessedConfig);
```

Comment:
The EnvironmentPipelineValidator does not seem to check if pipeline is a part of another environment. A pipeline can only belong to one environment. Can you please confirm or point out to where this validation happens?